### PR TITLE
support trace replay.

### PR DESCRIPTION
**Pull request by momo609**
## Motivation
follow by [vllm/46701](https://github.com/vllm-project/vllm/pull/46701) add support for forcing the decoder to emit a specified token sequence through trace_decode_token_ids. This is useful for request replay and performance analysis while preserving the model’s original probability distribution and returning the logprob of each forced token.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/srt/sampling/sampling_params.py [hunk 1]
```diff
@@ -152,6 +152,9 @@ class SamplingParams(msgspec.Struct, kw_only=True, array_like=True):
     logit_bias: Optional[Dict[str, float]] = None
     sampling_seed: Optional[int] = None
     custom_params: Optional[Dict[str, CustomParamValue]] = None
+    # Force the decoder to emit this token sequence while preserving the model's
+    # original sampling distribution for returned logprobs.
+    trace_decode_token_ids: Optional[List[int]] = None
```
Comment (Ronald1995) [inline python/sglang/srt/sampling/sampling_params.py:157]: we should also add this filed for rust frontend.

File: test/registered/unit/sampling/test_sampling_batch_info.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import torch
 
 from sglang.srt.constrained.base_grammar_backend import GrammarMask
+from sglang.srt.layers.sampler import apply_trace_decode_tokens
```
Comment (Ronald1995) [inline test/registered/unit/sampling/test_sampling_batch_info.py:15]: please add a e2e case to cover trace replay feature.

File: python/sglang/srt/sampling/sampling_params.py [hunk 3]
```diff
@@ -277,6 +280,35 @@ def verify(self, vocab_size):
...
+                token_id
+                for token_id in self.trace_decode_token_ids
+                if token_id >= vocab_size
+            ]
+            if invalid_token_ids:
+                raise ValueError(
+                    "trace_decode_token_ids contains out-of-vocabulary token ids: "
+                    f"{invalid_token_ids[:8]}"
+                )
+
```
Comment (Ronald1995) [inline python/sglang/srt/sampling/sampling_params.py:311]: please extract a method for this validation, what's more,  it seems like trace replay is also conflicted with 

```python
max_thinking_tokens
--enable-strict-thinking
ThinkingBudgetLogitProcessor
ReasonerGrammarObject
```

File: python/sglang/srt/layers/sampler.py [hunk 4]
```diff
@@ -76,6 +72,42 @@
...
+        or trace_token_lens is None
+        or trace_prompt_lens is None
+        or trace_token_ids.shape[1] == 0
+    ):
+        return sampled_token_ids
+
+    original_shape = sampled_token_ids.shape
+    sampled_token_ids = sampled_token_ids.reshape(-1)
+    positions = positions.to(torch.int64).reshape(-1)
+    trace_step = positions - trace_prompt_lens + 1
```
Comment (Ronald1995) [inline python/sglang/srt/layers/sampler.py:95]: it's better to calculate trace_step in SamplingBatchInfo like this 

```python

trace_decode_steps = torch.tensor(
    [
        len(req.output_ids)
        if req.sampling_params.trace_decode_token_ids is not None
        else 0
        for req in batch.reqs
    ]
)
```

**Review by Ronald1995 — CHANGES_REQUESTED** [review]
